### ISSUE TITLE
acPostProcForPut fires off also on replication

### ISSUE THREAD

**Issue by ilarik** [irods/irods#4614]
### iRODS Version, OS and Version
iRODS 4.2.6, CentOS 7.6.1810 (vagrant) and CentOS 7.7.1908 (production)

### What did you try to do?
When replicating objects from one resource to another, the rule engine seems to execute rule `acPostProcForPut` also on objects that get replicated (not only when being ingested).

### Expected behavior
No launch of `acPostProcForPut` at replication, only perhaps `acPostProcForRepl` as this can be an issue for rules meant to be applied to an entire data object, not to one replica.

### Observed behavior (including steps to reproduce, if applicable)
Can be reproduced using `irods-provisioner`:
```
$ git clone https://github.com/KTH-PDC/irods-provisioner.git 
$ cd irods-provisioner 
$ vagrant up
$ ANSIBLE_HOST_KEY_CHECKING=False ansible-playbook -b -i hosts-test irods-cluster.yml
```
After the test cluster is complete, some additions:
```
$ vagrant ssh rodsmaster1
[vagrant@rodsmaster1 ~]$ sudo su - irods
-bash-4.2$ cat >> /etc/irods/core-override.re 

acPostProcForPut
{
    writeLine("serverLog", "acPostProcForPut, objPath = '$objPath'");

    delay("<PLUSET>0m</PLUSET>")
    {
	writeLine("serverLog", "msiSysChksumDataObj, objpath = '$objPath'");
	msiSysChksumDataObj;
    }
}
^D
-bash-4.2$ iadmin mkresc randomResc random
-bash-4.2$ for i in {1..3}; do iadmin addchildtoresc randomResc rodsresc${i}_resc; done
```
Now testing against the VM setup:
```
$ iput -rv -R demoResc README.md 
   README.md                       0.002 MB | 0.022 sec | 0 thr |  0.081 MB/s
$ irepl -v -S demoResc -R randomResc README.md 
   README.md                       0.002 MB | 0.175 sec | 0 thr |  0.010 MB/s
```
in `rodsLog` at the VM we can observe:
```
Dec  3 17:14:52 pid:12898 NOTICE: writeLine: inString = acPostProcForPut, objPath = '/tempZone/home/rods/README.md'
Dec  3 17:15:03 pid:12902 NOTICE: writeLine: inString = msiSysChksumDataObj, objpath = '/tempZone/home/rods/README.md'
Dec  3 17:15:58 pid:12983 NOTICE: writeLine: inString = acPostProcForPut, objPath = '/tempZone/home/rods/README.md'
Dec  3 17:16:03 pid:12989 NOTICE: writeLine: inString = msiSysChksumDataObj, objpath = '/tempZone/home/rods/README.md'
```

**Comment by trel** [comment]
Yes, we know the static policy enforcement points fire at extra times - the dynamic policy enforcement points are enforced by the architecture to be fired correctly, per operation.  They are recommended over any static PEPs.

Will leave this open - but not sure how to fix at this time.

**Comment by ilarik** [comment]
Hi, any plans for a fix for this one, in any upcoming releases?

**Comment by trel** [comment]
Not currently.  If we change/fix where the static PEPs fire, we could break existing deployed policy.

Our current thinking is to provide a robust replacement and migration documentation to the dynamic PEPs for anyone still using the static PEPs.

Eventually, we would deprecate and retire the static PEPs.

**Comment by alanking** [comment]
I would like to offer an explanation for this behavior, for posterity.

First of all, I think we are standing by our thinking that dynamic PEPs will continue to be our recommendation as we are trying to sunset static PEPs. The equivalent dynamic PEP for the static PEP `acPostProcForPut` is `pep_api_data_obj_put_post`. Please notice that other operations such as rsync, copy, replication, and phymv use different dynamic PEPs, so make sure operations of concern in your deployment are covered when implementing. See https://docs.irods.org/4.2.8/plugins/dynamic_policy_enforcement_points/ for more information.

Now for the issue being described by @ilarik in this issue... the behavior broke in this change (authored, of course, by yours truly) (see #4099): https://github.com/irods/irods/commit/3b85607cf9b1e4d2eee6cbb4465f9b50a4ef3ef8

`acPostProcForPut`, unfortunately, is not fired exclusively for put operations. It is also called from `rsDataObjClose` under the following conditions (taken from 4.1.10, but it is identical in 4.2.8):
https://github.com/irods/irods/blob/bb48cd38c4a543ad8bf45082fa06b15290378b50/iRODS/server/api/src/rsDataObjClose.cpp#L189-L208

If the `openType` is set for the L1 descriptor of the object being closed and it meets the conditions described in the logic above, `acPostProcForPut` will be invoked.

Historically, `rsDataObjRepl` has not used `rsDataObjOpen` like most of the other APIs and has instead hand-crafted its own L1 descriptor in `dataObjOpenForRepl`. Part of this process was to set the `oprType` to `REPLICATE_DEST` (in the case of a destination for replication) but the `openType` was excluded. This allowed replications to snake around the logic in `rsDataObjClose` and avoid invoking `acPostProcForPut` erroneously.

In the change linked above, `openType` was added to the logic for `rsDataObjRepl` so that `file_modified` would not be invoked erroneously for replications (as the `openType` is used in the determination as to whether `file_modified` should be invoked). As a result, replications (whether on create or on overwrite) now meet the criteria for `acPostProcForPut` in `rsDataObjClose`'s handling of the static post-PEPs.

This explains why `acPostProcForRepl` no longer fires on creates and why `acPostProcForCreate` and `acPostProcForPut` are now firing instead. I also noticed strange behavior with this and other static PEPs (see https://github.com/irods/irods/issues/4543 and comments therein). The reason this has laid largely undetected within the consortium, I expect, is due to insufficient testing around static PEPs. Therefore, we recommend avoiding static PEPs and using dynamic PEPs instead.
